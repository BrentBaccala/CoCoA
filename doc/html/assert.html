<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML>
<HEAD>
<META NAME="generator" CONTENT="http://txt2tags.org">
<LINK REL="stylesheet" TYPE="text/css" HREF="cocoalib-doc.css">
<TITLE>      assert</TITLE>
</HEAD><BODY BGCOLOR="white" TEXT="black">
<CENTER>
<H1>      assert</H1>
<FONT SIZE="4"><I>      &copy;  2007 John Abbott</I></FONT><BR>
<FONT SIZE="4">      GNU Free Documentation License, Version 1.2</FONT>
</CENTER>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>

    <UL>
    <LI><A HREF="#examples">Examples</A>
    <LI><A HREF="#toc2">User documentation for files assert.H and assert.C</A>
      <UL>
      <LI><A HREF="#toc3">Debugging</A>
      </UL>
    <LI><A HREF="#toc4">Maintainer documentation for files assert.H and assert.C</A>
    <LI><A HREF="#toc5">Bugs, Shortcomings, and other ideas</A>
    </UL>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>
<P>
      <center><a href="index.html">CoCoALib Documentation Index</a></center>
</P>

<A NAME="examples"></A>
<H2>Examples</H2>

<UL>
<LI><A HREF="../../examples/index.html#ex-PolyIterator2.C">ex-PolyIterator2.C</A>
</UL>

<A NAME="toc2"></A>
<H2>User documentation for files assert.H and assert.C</H2>

<P>
The only part of <CODE>assert.H</CODE> which a normal CoCoALib user might find useful
is the <CODE>CoCoA_ASSERT</CODE> macro: it is intended to be a debugging aid.
</P>
<P>
The <CODE>CoCoA_ASSERT</CODE> macro does absolutely nothing (not even evaluating its
argument) unless the compilation flag <CODE>CoCoA_DEBUG</CODE> is set.  If that flag
is set then the macro evaluates its argument to a boolean result which
is then tested: if the result is true nothing further happens, if the
result is false then the function <CODE>CoCoA::AssertionFailed</CODE> is called with
some arguments indicating which <CODE>CoCoA_ASSERT</CODE> macro call obtained the false
value.  The <CODE>AssertionFailed</CODE> function prints out an informative message on
<CODE>std::cerr</CODE> and then throws a <CODE>CoCoA::ERR::AssertFail</CODE> exception.
</P>
<P>
The file <CODE>assert.H</CODE> contains the following:
-  definition of the <CODE>CoCoA_ASSERT</CODE> macro to aid debugging, and the
     related function <CODE>AssertionFailed</CODE>
</P>

<A NAME="toc3"></A>
<H3>Debugging</H3>

<P>
During debugging, a debugger can be used to intercept calls to the
function <CODE>CoCoA::AssertionFailed</CODE> which will stop the program just before
throwing the <CODE>CoCoA::ERR::AssertFail</CODE> exception.  This should enable one
to find more easily the cause of the problem.
</P>
<P>
For example, in <B>gdb</B> type
</P>

<PRE>
  break CoCoA::AssertionFailed
</PRE>

<P>
and then go <CODE>up</CODE> (perhaps repeatedly) to the offending line.
</P>

<A NAME="toc4"></A>
<H2>Maintainer documentation for files assert.H and assert.C</H2>

<P>
The macro name <CODE>CoCoA_ASSERT</CODE> is rather cumbersome, but must contain the
prefix <CODE>CoCoA_</CODE> since macro names cannot be placed in C++ namespaces.
The two definitions of the macro (debugging and non-debugging cases)
both look rather clumsy, but are done that way so that the macro expands
into an expression which is syntactically a simple command.  The
definition for the non-debugging case I took from <CODE>/usr/include/assert.h</CODE>;
I do not recall where I got the definition for the debugging case, but
the definition in <CODE>/usr/include/assert.h</CODE> looked to be gcc specific.
</P>
<P>
The purpose of the procedure <CODE>AssertionFailed</CODE> is explained above in the
user documentation (to facilitate interception of failed assertions).  The
procedure never returns; instead it throws a CoCoALib exception with code
<CODE>ERR::AssertFail</CODE>.  Before throwing the exception it prints out a message
on <CODE>std::cerr</CODE> summarising what the assertion was, and where it was.
Note the non-standard way of throwing the CoCoA exception: this allows the
<CODE>ErrorInfo</CODE> object to refer to the file and line where <CODE>CoCoA_ASSERT</CODE> was
called (rather then to the line in <CODE>assert.C</CODE> where <CODE>CoCoA_ERROR</CODE> is called).
The entire printed message is assembled into an <CODE>ostringstream</CODE> before being
printed to ensure exception safety: either the whole message is printed or
none of it is, since the printing step is an atomic operation.
</P>

<A NAME="toc5"></A>
<H2>Bugs, Shortcomings, and other ideas</H2>

<P>
Is the exception safe implementation of <CODE>AssertionFailed</CODE> excessive?
</P>
<P>
You have to use explicitly <CODE>#ifdef CoCoA_DEBUG</CODE> if you want to have a
loop or any other non-trivial piece of code executed only when debugging
it turned on.
</P>
<P>
The following (simplified but real) code excerpt is mildly problematic:
</P>

<PRE>
  {
    bool OK = ....;
    CoCoA_ASSERT(OK);
  }
</PRE>

<P>
When compiled without debugging (<I>i.e.</I> <CODE>CoCoA_DEBUG</CODE> is zero) the compiler
(gcc-3) complains that the variable <CODE>OK</CODE> is unusued.  It does not appear
to be possible to make the macro "depend on its argument" in the
non-debugging case without incurring the run-time cost of evaluating the
argument (if the argument is just a variable the cost is negligible, but
if it is a more complex expression then the cost could be considerable).
The solution adopted was to modify the calling code like this:
</P>

<PRE>
  {
    bool OK;
    OK = ....;
    CoCoA_ASSERT(OK);
  }
</PRE>

<P>
Note that the apparently simpler code below <B>will not work</B> if the elided
code (<I>i.e.</I> the <CODE>....</CODE>) has a side effect since the elided code will not
be called at all in the non-debugging case:
</P>

<PRE>
  {
    CoCoA_ASSERT(....);
  }
</PRE>

<P>
<B>POSSIBLE SOLUTION</B>: maybe <CODE>CoCoA_ASSERT</CODE> could compute <CODE>sizeof(...)</CODE> in the
non-debugging case -- this should avoid evaluation of the argument, and
will compile away to nothing.
</P>

<!-- html code generated by txt2tags 2.6 (http://txt2tags.org) -->
<!-- cmdline: txt2tags -t html -o html/assert.html -i txt/assert.txt -->
</BODY></HTML>
