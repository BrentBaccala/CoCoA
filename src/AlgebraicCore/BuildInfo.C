//   Copyright (c)  2005,2007,2012  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/config.H"

#include <limits>
#include <iostream>
using std::ostream;
using std::endl;
#include <string>
using std::string;

namespace CoCoA
{

  namespace BuildInfo
  {
    // Build info is actually passed in via three preprocessor variables:
    // COCOA_VERSION, COCOA_CXX, COCOA_CXXFLAGS -- all three are quoted strings.

    const std::string& version()
    {
      static const string info(COCOA_VERSION);
      return info;
    }

    const std::string& compiler()
    {
      static const string info(COCOA_CXX);
      return info;
    }

    const std::string& CompilationFlags()
    {
      static const string info(COCOA_CXXFLAGS);
      return info;
    }

    const std::string& CompilationDefines()
    {
      static const string info(COCOA_CXXFLAGS_DEFINES);
      return info;
    }


    void PrintAll(std::ostream& out)
    {
    // Below we use string literal juxtaposition, so that the build
    // info can readily be extracted from libcocoa.a by the command
    // strings libcocoa.a | egrep "CoCoA::BuildInfo"
      out << endl
          << "CoCoA::BuildInfo Summary of build information for CoCoALib:" << endl
          << "CoCoA::BuildInfo CoCoALib Version: " COCOA_VERSION << endl
          << "CoCoA::BuildInfo Compiler: " COCOA_CXX << endl
          << "CoCoA::BuildInfo Compilation Flags: " COCOA_CXXFLAGS << endl
          << "CoCoA::BuildInfo Compilation Flags Defines: " COCOA_CXXFLAGS_DEFINES << endl
          << "CoCoA::BuildInfo NumBits int = " << std::numeric_limits<unsigned int>::digits << endl
          << "CoCoA::BuildInfo NumBits long = " << std::numeric_limits<unsigned long>::digits << endl
          << "CoCoA::BuildInfo NumBits SmallExponent_t = " << std::numeric_limits<SmallExponent_t>::digits << endl
          << endl;
    }

  } // end of namespace CoCoA::BuildInfo

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/BuildInfo.C,v 1.8 2012/10/24 12:10:45 abbott Exp $
// $Log: BuildInfo.C,v $
// Revision 1.8  2012/10/24 12:10:45  abbott
// Revised PrintAll: now prints also NumBits of SmallExponent_t
//
// Revision 1.7  2012/10/02 11:45:09  bigatti
// -- added CompilationDefines
//
// Revision 1.6  2012/10/02 10:36:12  abbott
// Revised interface to BuildInfo information strings.
// Several consequential changes.
//
// Revision 1.5  2012/06/20 12:26:26  bigatti
// -- now exporting also compiler and compilation flags
//
// Revision 1.4  2011/03/11 17:38:21  bigatti
// -- added missing include
//
// Revision 1.3  2011/03/11 17:18:55  bigatti
// -- added info on int and long size
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.4  2007/02/10 18:44:03  cocoa
// Added "const" twice to each test and example.
// Eliminated dependency on io.H in several files.
// Improved BuildInfo, and added an example about how to use it.
// Some other minor cleaning.
//
// Revision 1.3  2007/02/08 22:34:22  cocoa
// Changed BuildInfo: only BuildInfo version is publicly visible.
// Only BuildInfo needs complicated compile-time flags, so several changes
// to Makefiles etc.  Added a new example: ex-BuildInfo.C
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.2  2005/10/12 15:52:09  cocoa
// Completed test-RingFp1 and corrected/cleaned the SmallFp*
// and RingFp* files.
//
// Some minor tidying elsewhere.
//
// Revision 1.1  2005/10/06 16:36:42  cocoa
// Added the capability find out build information at run-time.
// The Makefiles should be a little tidier too.
//
