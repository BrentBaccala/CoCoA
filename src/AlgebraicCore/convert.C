//   Copyright (c)  2007,2009  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.

#include "CoCoA/convert.H"
#include "CoCoA/BigRat.H"
#include "CoCoA/IntOperations.H"
#include "CoCoA/ring.H"

#include <limits>
using std::numeric_limits;
#include <cmath>
using std::ldexp;
using std::floor;

namespace CoCoA
{

  //-------------------------------------------------------
  // BigInt

  bool IsConvertible(long& lhs, const BigInt& src)
  {
    if (!mpz_fits_slong_p(mpzref(src))) return false;
    lhs = mpz_get_si(mpzref(src));
    return true;
  }

  bool IsConvertible(int& lhs, const BigInt& src)
  {
    if (!mpz_fits_sint_p(mpzref(src))) return false;
    lhs = mpz_get_si(mpzref(src));
    return true;
  }

  bool IsConvertible(unsigned long& lhs, const BigInt& src)
  {
    if (!mpz_fits_ulong_p(mpzref(src))) return false;
    lhs = mpz_get_ui(mpzref(src));
    return true;
  }

  bool IsConvertible(unsigned int& lhs, const BigInt& src)
  {
    if (!mpz_fits_uint_p(mpzref(src))) return false;
    lhs = mpz_get_ui(mpzref(src));
    return true;
  }


  //-------------------------------------------------------
  // BigRat

  bool IsConvertible(long& lhs, const BigRat& src)
  {
    // Simple rather than fast -- makes unnecessary copy of num(src)
    return IsOneDen(src) && IsConvertible(lhs, num(src));
  }

  bool IsConvertible(int& lhs, const BigRat& src)
  {
    // Simple rather than fast -- makes unnecessary copy of num(src)
    return IsOneDen(src) && IsConvertible(lhs, num(src));
  }

  bool IsConvertible(unsigned long& lhs, const BigRat& src)
  {
    // Simple rather than fast -- makes unnecessary copy of num(src)
    return IsOneDen(src) && IsConvertible(lhs, num(src));
  }

  bool IsConvertible(unsigned int& lhs, const BigRat& src)
  {
    // Simple rather than fast -- makes unnecessary copy of num(src)
    return IsOneDen(src) && IsConvertible(lhs, num(src));
  }


  //-------------------------------------------------------
  // double with long, BigInt, BigRat

  bool IsConvertible(long& n, double z)
  {
    //??? BUG should handle infinity and Nan???
    if (z != std::floor(z)) return false;
    if (z < numeric_limits<long>::min() || z > numeric_limits<long>::max()) return false;
    n = static_cast<long>(z); // already checked that z is integer and in range.
    return true;
  }


  // NB ALL sufficiently large doubles are considered to be integers
  bool IsConvertible(BigInt& N, double z)
  {
    //??? BUG should handle infinity and Nan???
    if (z != std::floor(z)) return false;
    mpz_set_d(mpzref(N), z);
    return true;
  }




  bool IsConvertible(BigRat& Q, double z)
  {
    //??? BUG should handle infinity and Nan???
    if (z == 0) { Q = 0; return true; }
    mpq_set_d(mpqref(Q), z);  // BUG Don't know what happens if this fails.
    return true;
  }


  // Sets z = N if possible.
  bool IsConvertible(double& z, const BigInt& N)
  {
    if (IsZero(N)) { z = 0.0; return true; }
    // I prefer to do this manually because mpz_get_d does not give
    // good guarantees about what happens when overflow occurs.
    long ExpN;
    double MantissaN = mpz_get_d_2exp(&ExpN, mpzref(N));
    if (ExpN > numeric_limits<double>::max_exponent)
    {
      return false;
    }
    z = ldexp(MantissaN, ExpN);
    return true;
  }


  // Sets z = num/den if possible.
  // Underflow sets z to 0 and returns true.
  // Overflow returns false (and does not change z).
  bool IsConvertible(double& z, const BigRat& Q)
  {
    if (IsZero(Q)) { z = 0.0; return true; }
    // I prefer to do this manually rather than use mpq_get_d because
    // num and den may not be coprime, and the cost of computing the
    // gcd could be high.  Also mpq_get_d does not give nice guarantees
    // about what happens in the case of over-/under-flow.
    long Nexp;
    double N = mpz_get_d_2exp(&Nexp,mpq_numref(mpqref(Q)));
    long Dexp;
    double D = mpz_get_d_2exp(&Dexp, mpq_denref(mpqref(Q)));
    long ExpAns = Nexp-Dexp;
    double MantissaAns = N/D;
    if (MantissaAns >= 1 || MantissaAns <= -1)
    { MantissaAns /= 2; ++ExpAns; }

    // I believe this is the right interpretation of the exponent limits...
    if (ExpAns < numeric_limits<double>::min_exponent)
    {
      // We have underflow, so convert to 0 and indicate "success"
      z = 0.0;
      return true;
    }

    if (ExpAns > numeric_limits<double>::max_exponent)
    {
      // We have overflow, so indicate "failure".
      return false;
    }

    z = ldexp(MantissaAns, ExpAns); // Cannot overflow or underflow.
    return true;
  }


  //-------------------------------------------------------
  // RingElem to long, BigInt, BigRat

  bool IsConvertible(long& n, ConstRefRingElem x)
  {
    BigInt N;
    return (IsConvertible(N, x) && IsConvertible(n, N));
  }

  bool IsConvertible(BigInt& N, ConstRefRingElem x)
  {
    return IsInteger(N, x);
  }

  bool IsConvertible(BigRat& N, ConstRefRingElem x)
  {
    return IsRational(N, x);
  }


} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/convert.C,v 1.23 2014/04/02 15:55:45 abbott Exp $
// $Log: convert.C,v $
// Revision 1.23  2014/04/02 15:55:45  abbott
// Summary: Reorganized file contents (more sensible order now, I hope)
// Author: JAA
//
// Revision 1.22  2013/03/26 14:56:06  abbott
// Updated the conversion fns (in ptic removed procedure "convert");
// numerous consequential changes.
//
// Revision 1.21  2013/02/15 16:32:44  abbott
// Removed IsInRange (moved to MachineInt).
// Added IsConvertible(BigInt, RingElem) and IsConvertible(BigRat, RingElem).
//
// Revision 1.20  2012/05/28 09:18:20  abbott
// Created IntOperations which gathers together all operations on
// integers (both big and small).  Many consequential changes.
//
// Revision 1.19  2011/11/09 15:35:00  bigatti
// -- moved include MachineInt into convert.H
//
// Revision 1.18  2011/11/09 14:29:37  bigatti
// -- renamed MachineInteger --> MachineInt
//
// Revision 1.17  2011/08/27 20:50:47  abbott
// Added include directive for MachineInt.
//
// Revision 1.16  2011/08/24 10:32:04  bigatti
// -- renamed QQ --> BigRat
// -- sorted #include
//
// Revision 1.15  2011/08/14 15:52:16  abbott
// Changed ZZ into BigInt (phase 1: just the library sources).
//
// Revision 1.14  2011/08/12 16:08:32  abbott
// Hastily added conversion fns for BigInt (like those for ZZ).
//
// Revision 1.13  2011/03/22 20:26:12  abbott
// Added  IsConvertible(long&, double).
// Activated commented out ConvertTo template.
// Checking in because some other files need the new code.
//
// Revision 1.12  2010/04/23 13:21:45  abbott
// Conversion from string to ZZ now defaults to decimal (and not the quirky C standard).
//
// Revision 1.11  2010/03/22 11:51:37  abbott
// Simplified convert fn from string to BigRat -- now uses GMP's implementation.
//
// Revision 1.10  2009/12/23 22:27:28  abbott
// Added conversions from BigRat to machine integers.
//
// Revision 1.9  2009/12/23 18:53:51  abbott
// Major change to conversion functions:
//   convert(..) is now a procedure instead of a function
//   IsConvertible(..) replaces the former convert(..) function
//   Added new NumericCast conversion function (placeholder for BOOST feature)
//   Consequent changes in code which uses these features.
//
// Revision 1.8  2009/12/11 11:46:32  abbott
// Changed fn  convert  into  IsConvertible.
// Added template procedure  convert.
// New version because change is not backward compatible.
//
// Revision 1.7  2009/07/08 12:26:53  abbott
// Added floor and ceil functions for BigRats.
// Added example program for BigRats.
// Minor correction to convert.C; minor cleaning to ex-ZZ1.C
//
// Revision 1.6  2009/07/02 16:32:11  abbott
// Consequential changes stemming from new class BigRat, and modified interface to the member
// function RingBase::myIsRational.  Also some new conversion functions.
//
// Revision 1.5  2008/10/07 15:45:22  abbott
// Changed ErrorInfo objects so they include the name of their own error ID.
// Changed catch statements to catch const objects.
// Removed calls to the member fn which accessed the error ID member of an
// ErrorInfo; now you simply compare directly with the error ID (makes the
// code easier to read).
//
// Revision 1.4  2007/10/30 17:14:07  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.3  2007/05/22 22:51:39  abbott
// Changed name of fn ndigits to NumDigits.
// Changed return type of exponent and NumDigits.
// Changed some exceptions from nonstandard to the appropriate standard one.
//
// Revision 1.2  2007/03/28 15:34:03  abbott
// Underflow during conversion to a double no longer causes "failure" but now
// converts the value to zero and yields "success".
//
// Revision 1.1  2007/03/23 18:38:42  abbott
// Separated the "convert" functions (and CheckedCast) into their own files.
// Many consequential changes.  Also corrected conversion to doubles.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.3  2007/03/03 15:24:39  cocoa
// Changed 2006 to 2007.
//
// Revision 1.2  2006/10/06 14:04:14  cocoa
// Corrected position of #ifndef in header files.
// Separated CoCoA_ASSERT into assert.H from config.H;
// many minor consequential changes (have to #include assert.H).
// A little tidying of #include directives (esp. in Max's code).
//
