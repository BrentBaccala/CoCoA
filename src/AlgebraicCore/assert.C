//   Copyright (c)  2005  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/assert.H"
#include "CoCoA/error.H"

#include <iostream>
using std::cerr;
using std::endl;
#include <sstream>
using std::ostringstream;


namespace CoCoA
{

  void AssertionFailed(const char* const cond, const char* const file, unsigned long line)
  {
    ostringstream message;
    message << endl
            << "===========================================================================" << endl
            << "===========================================================================" << endl
            << "==== CoCoA Assertion failed: [[" << cond << "]]" << endl
            << "==== File: " << file << endl
            << "==== Line: " << line << endl
            << "===========================================================================" << endl
            << "===========================================================================" << endl
            << endl;
    cerr << message.str();

    // Throw a CoCoA error in an unusual way... (so that the "right" file and line no. info appear).
    ThrowError(ErrorInfo(ERR::AssertFail, "AssertionFailed", file, line));
  }

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/assert.C,v 1.3 2010/12/17 16:11:06 abbott Exp $
// $Log: assert.C,v $
// Revision 1.3  2010/12/17 16:11:06  abbott
// Output related to assertions is now on standard C++ streams (rather
// than GlobalErrput, etc).
//
// Revision 1.2  2007/10/30 17:14:07  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.1  2006/10/06 14:04:57  cocoa
// The new assert header and implementation files.
// A new test.
//
//
