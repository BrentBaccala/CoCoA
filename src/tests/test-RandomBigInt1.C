//   Copyright (c)  2010  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/GlobalManager.H"
#include "CoCoA/IntOperations.H"
#include "CoCoA/random.H"
#include "CoCoA/assert.H"  // for using TEST_ASSERT
#include "CoCoA/convert.H"
#include "CoCoA/error.H"


#include <iostream>
using std::cout;
using std::cerr;
using std::endl;
#include <vector>
using std::vector;

using namespace CoCoA;

// Handy macro for making assertions.
#define TEST_ASSERT(cond)   do { if (!(cond)) CoCoA::AssertionFailed(#cond, __FILE__, __LINE__); } while(0)

void program()
{
  GlobalManager CoCoAFoundations;

  const int N = 10;
  const BigInt midpoint = power(3,50);
  RandomSeqBigInt RndBigInt(midpoint-N, midpoint+N);
  cout << "RndBigInt=" << RndBigInt << endl;

  const int ExpectedFreq = 256; // should be larger than 100ish
  vector<int> histo(2*N+1);
  for (int i=0; i < ExpectedFreq*(2*N+1); ++i)
  {
    const BigInt& rnd = NextValue(RndBigInt) - midpoint;
    long delta;
    TEST_ASSERT(IsConvertible(delta, rnd));
    TEST_ASSERT(delta >= -N && delta <= N);
    ++histo[N+delta];
  }

  // Check histogram: each entry is > 0, and greatest and least freqs are within a factor of 2
  int MostFreq = histo[0];
  int LeastFreq = histo[0];
  for (int i=0; i < 2*N+1; ++i)
  {
    TEST_ASSERT(histo[i] != 0);
    if (histo[i] > MostFreq) { MostFreq = histo[i]; continue; }
    if (histo[i] < LeastFreq) { LeastFreq = histo[i]; }
  }
  const double ratio = double(MostFreq)/LeastFreq;
  TEST_ASSERT(ratio < 2.0);

  cout << "RndBigInt=" << RndBigInt << endl;
}


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  BuildInfo::PrintAll(cerr);
  return 1;
}
