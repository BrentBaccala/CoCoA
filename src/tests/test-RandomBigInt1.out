RndBigInt=RandomSeqBigInt(lwb=717897987691852588770239, upb=717897987691852588770259, seed=0, counter=0)
RndBigInt=RandomSeqBigInt(lwb=717897987691852588770239, upb=717897987691852588770259, seed=0, counter=5376)
